### PR TITLE
LockedGold Foundry tests

### LINKED ISSUE #10799: test/governance/voting/lockedgold.ts

### PR DESCRIPTION
**Pull request by pahor167**
### Description

LockedGold Foundry tests migrated from Truffle

Closes #10799

### CODE REVIEW

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 1]
```diff
@@ -0,0 +1,2890 @@
...
+import "../../../contracts/common/test/MockGoldToken.sol";
+import "../../../contracts/governance/LockedGold.sol";
+import "../../../contracts/governance/ReleaseGold.sol";
+import "../../../contracts/governance/Election.sol";
+import "../../../contracts/stability/test/MockStableToken.sol";
+import "../../../contracts/governance/test/MockElection.sol";
+import "../../../contracts/governance/test/MockGovernance.sol";
+import "../../../contracts/governance/test/MockValidators.sol";
+
+contract LockedGoldFoundryTest is Test {
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
contract LockedGoldTest is Test {
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 2]
```diff
@@ -0,0 +1,2890 @@
...
+    (uint256 fraction, uint256 currentAmount) = lockedGold.getDelegatorDelegateeInfo(
+      _delegator,
+      delegatee
+    );
+    assertEq(FixidityLib.wrap(fraction * 100).fromFixed(), percent, "fraction incorrect");
+    assertEq(currentAmount, amount, "amount incorrect");
+  }
+}
+
+contract Initialize is LockedGoldFoundryTest {
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I would suggest including the name of the contract in all the sub-contracts representing a `describe` scope. In this instance, from `Initialize` to `LockedGoldInitialize`

This would allow to more effectively use the test filtering function that comes with forge.

For instance, if we wanted to only test this function, you couldnt filter using `--match-contract Initialize`

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 3]
```diff
@@ -0,0 +1,2890 @@
...
+    lockedGold.lock.value(value)();
+    assertEq(lockedGold.getAccountNonvotingLockedGold(address(this)), value);
+  }
+
+  function test_ShouldIncreaseTheAccountTOtalLockedGoldBalance() public {
+    lockedGold.lock.value(value)();
+    assertEq(lockedGold.getAccountTotalLockedGold(address(this)), value);
+  }
+
+  function test_ShouldIncreaseTheNonvotingLockedGOldBalance() public {
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  function test_ShouldIncreaseTheNonvotingLockedGoldBalance() public {
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 4]
```diff
@@ -0,0 +1,2890 @@
...
+    public
+  {
+    validators.setAccountLockedGoldRequirement(address(this), balanceRequirement);
+    vm.expectRevert(
+      "Either account doesn't have enough locked Celo or locked Celo is being used for voting."
+    );
+    lockedGold.unlock(value);
+  }
+
+  function test_ShouldSucceed_WhenTheCorrectTimeIsEarlierThanThRequirementTimeButRequestingCeloWithoutBalanceRequirement_WhenThereIsBalanceRequirement()
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  function test_ShouldSucceed_WhenTheCorrectTimeIsEarlierThanTheRequirementTimeButRequestingCeloWithoutBalanceRequirement_WhenThereIsBalanceRequirement()
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 5]
```diff
@@ -0,0 +1,2890 @@
...
+  function test_ShouldEmitGoldUnlockedEvent_WhenTheAccountIsRequestingOnlyNonVotingGold() public {
+    governance.setVoting(address(this));
+    governance.setTotalVotes(address(this), votingGold);
+
+    vm.expectEmit(true, true, true, true);
+    emit GoldUnlocked(address(this), nonVotingGold, availabilityTime);
+    lockedGold.unlock(nonVotingGold);
+  }
+
+  function test_ShouldRevert_WhenTheCorrectTimeIsEarlierThanThRequirementTime_WhenThereIsBalanceRequirement()
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  function test_ShouldRevert_WhenTheCorrectTimeIsEarlierThanTheRequirementTime_WhenThereIsBalanceRequirement()
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 6]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+
+  function test_ShouldRevertWhenAccountDoesNotExist() public {
+    vm.expectRevert("Must first register address with Account.createAccount");
+    vm.prank(randomAddress);
+    lockedGold.lock();
+  }
+}
+
+contract Unlock is LockedGoldFoundryTest {
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: The different scenarios for unlocking does not seem to all be captured here.

e.g.: It's not clear which test handles the scenario "when there are no balance requirements && when the account is not voting in governance" vs "when there are no balance requirements && when the account is voting in governance"

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 7]
```diff
@@ -0,0 +1,2890 @@
...
+    vm.prank(delegator);
+    lockedGold.lock.value(value)();
+    lockedGold.updateDelegatedAmount(delegator, delegatee);
+
+    assertEq(lockedGold.getAccountTotalLockedGold(delegator), value * 2);
+    assertEq(lockedGold.getAccountTotalGovernanceVotingPower(delegatee), delegatedAmount * 2);
+    assertDelegatorDelegateeAmounts(delegator, delegatee, delegatedPercent, delegatedAmount * 2);
+  }
+
+  function test_ShouldReturnCorrectValueWhenLockedAndDelegated_WhenDelegatorLockedMoreGold_WhenVoteSigners()
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  function test_ShouldReturnCorrectValueWhenLockedAndDelegated_WhenDelegatorLockedMoreCelo_WhenVoteSigners()
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 8]
```diff
@@ -0,0 +1,2890 @@
...
+  function helper_WhenVoteSigners() public {
+    createAndAssertDelegatorDelegateeSigners(
+      delegator,
+      delegatee,
+      delegatorSignerPK,
+      delegateeSignerPK
+    );
+  }
+
+  function test_ShouldReturnCorrectValueWhenLockedAndDelegated_WhenDelegatorLockedMoreGold()
```
Comment (soloseng) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  function test_ShouldReturnCorrectValueWhenLockedAndDelegated_WhenDelegatorLockedMoreCelo()
```

**Review by soloseng — APPROVED** [review]
🚀

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 9]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+}
+
+contract Initialize is LockedGoldFoundryTest {
+  function setUp() public {
+    super.setUp();
+  }
+
+  function test_ShouldSetOwner() public {
+    assertEq(lockedGold.owner(), address(this));
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: To mock owner I usually define `address caller = address(this);` in the contract, this could even be added to this file I created: https://github.com/celo-org/celo-monorepo/pull/10773/files#diff-f7395554b886736da22e519de74623b3087a6c9239a351df77c8586cc6841c07

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 10]
```diff
@@ -0,0 +1,2890 @@
...
+
+  function test_ShouldEmitAGoldLockedEvent() public {
+    vm.expectEmit(true, true, true, true);
+    emit GoldLocked(address(this), value);
+    lockedGold.lock.value(value)();
+  }
+
+  function test_ShouldRevertWhenAccountDoesNotExist() public {
+    vm.expectRevert("Must first register address with Account.createAccount");
+    vm.prank(randomAddress);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:321]: Here I use `msg.sender`, so test consistently uses the same address.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 11]
```diff
@@ -0,0 +1,2890 @@
...
+    lockedGold.lock();
+  }
+}
+
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:331]: Should revert when user doesn't have enough balance?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 12]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+
+  function test_ShouldRevertWhenUnlockingGoldThatIsVotedWith() public {
+    governance.setVoting(address(this));
+    governance.setTotalVotes(address(this), votingGold);
+
+    vm.expectRevert("Not enough unlockable celo. Celo is locked in voting.");
+    lockedGold.unlock(value);
+  }
+
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:413]: should revert when unlocking more than locked amount?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 13]
```diff
@@ -0,0 +1,2890 @@
...
+  function test_ShouldEmitGoldUnlockedEvent() public {
+    vm.expectEmit(true, true, true, true);
+    emit GoldUnlocked(address(this), value, availabilityTime);
+    lockedGold.unlock(value);
+  }
+
+  function test_ShouldRevertWhenUnlockingGoldThatIsVotedWith() public {
+    governance.setVoting(address(this));
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Is there any reason we need this mock function and can not just simulate an actual vote?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 14]
```diff
@@ -0,0 +1,2890 @@
...
+    emit GoldUnlocked(address(this), nonVotingGold, availabilityTime);
+    lockedGold.unlock(nonVotingGold);
+  }
+
+  function test_ShouldRevert_WhenTheCorrectTimeIsEarlierThanTheRequirementTime_WhenThereIsBalanceRequirement()
+    public
+  {
+    validators.setAccountLockedGoldRequirement(address(this), balanceRequirement);
+    vm.expectRevert(
+      "Either account doesn't have enough locked Celo or locked Celo is being used for voting."
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:501]: somehow GitHub linter is breaking (at least for me) from here, but it doesn't seem be a syntax error.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 15]
```diff
@@ -0,0 +1,2890 @@
...
+    );
+    lockedGold.unlock(value);
+  }
+
+  function test_ShouldSucceed_WhenTheCorrectTimeIsEarlierThanTheRequirementTimeButRequestingCeloWithoutBalanceRequirement_WhenThereIsBalanceRequirement()
+    public
+  {
+    validators.setAccountLockedGoldRequirement(address(this), balanceRequirement);
+    lockedGold.unlock(value - balanceRequirement);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:510]: worth adding a check that the number is actually right?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 16]
```diff
@@ -0,0 +1,2890 @@
...
+
+  function test_ShouldRevertWhenUnlockingGoldThatIsVotedWith() public {
+    governance.setVoting(address(this));
+    governance.setTotalVotes(address(this), votingGold);
+
+    vm.expectRevert("Not enough unlockable celo. Celo is locked in voting.");
+    lockedGold.unlock(value);
+  }
+
+  function test_ShouldAddAPendingWithdrawal_WhenTheAccountIsRequestingOnlyNonVotingGold() public {
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
  // When the account is delegating
  function test_ShouldAddAPendingWithdrawal_WhenTheAccountIsRequestingOnlyNonVotingGold() public {
```

Comment (martinvol) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I'd add comments like that when the condition of the test change within the same class to make it easier to read.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 17]
```diff
@@ -0,0 +1,2890 @@
...
+    assertEq(lockedGold.totalDelegatedCelo(delegatee2), originallyDelegatedAmount / 2);
+  }
+
+  function test_ShouldCallRemoveDelegateVotesBecauseOfVotingForGovernance() public {
+    assertEq(governance.removeVotesCalledFor(delegatee), originallyDelegatedAmount / 2);
+    assertEq(governance.removeVotesCalledFor(delegatee2), originallyDelegatedAmount / 2);
+  }
+}
+
+contract UnlockDelegation3Delegatees is LockedGoldFoundryTest {
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]:
```suggestion
contract UnlockDelegationTo3Delegatees is LockedGoldFoundryTest {
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 18]
```diff
@@ -0,0 +1,2890 @@
...
+  function setUp() public {
+    super.setUp();
+    lockedGold.lock.value(pendingWithdrawalValue)();
+  }
+
+  function test_ShouldIncreaseTheAccountsNonVotingLockedGoldBalance_WhenRelockingValueEqualToTheValueOfThePendingWithdrawal_WhenPendingWithdrawalExists()
+    public
+  {
+    lockedGold.unlock(pendingWithdrawalValue);
+    lockedGold.relock(index, pendingWithdrawalValue);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I would be in favor of unifying all the functions that start with

```
lockedGold.unlock(pendingWithdrawalValue);
lockedGold.relock(index, pendingWithdrawalValue);
```

Comment (martinvol) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: or create a new setup with just this

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 19]
```diff
@@ -0,0 +1,2890 @@
...
+    );
+    assertEq(vals.length, 0);
+    assertEq(timestamps.length, 0);
+  }
+
+  function test_ShouldIncreaseTheAccountsNonVotingLockedGoldBalance_WhenRelockingValueLessThanValueOfThePendingWithdrawal_WhenPendingWithdrawalExists()
+    public
+  {
+    lockedGold.unlock(pendingWithdrawalValue);
+    lockedGold.relock(index, pendingWithdrawalValue - 1);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Ditto here with 
```
lockedGold.unlock(pendingWithdrawalValue);
lockedGold.relock(index, pendingWithdrawalValue - 1);
```

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 20]
```diff
@@ -0,0 +1,2890 @@
...
+contract AddSlasher is LockedGoldFoundryTest {
+  address downtimeSlasher = actor("DowntimeSlasher");
+
+  function setUp() public {
+    super.setUp();
+    registry.setAddressFor("DowntimeSlasher", downtimeSlasher);
+  }
+
+  function test_ShouldBeAbleToAddSlasherToWhitelist() public {
+    string memory slasherName = "DowntimeSlasher";
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: would move `slasherName = "DowntimeSlasher";`to the contract storage and use the variable in all the tests

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 21]
```diff
@@ -0,0 +1,2890 @@
...
+    public
+  {
+    uint256 voting = value / 2;
+    uint256 nonVoting = value - voting;
+    uint256 penalty = nonVoting;
+    uint256 reward = penalty / 2;
+    electionSlashTest.vote(group, voting, address(0), address(0));
+
+    helper_WhenAccountIsSlashedForAllOfItsLockedGold(penalty, reward);
+    assertEq(lockedGold.getAccountNonvotingLockedGold(address(this)), nonVoting - penalty);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Shouldn't we also check the voting gold?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 22]
```diff
@@ -0,0 +1,2890 @@
...
+    public
+  {
+    uint256 voting = value / 2;
+    uint256 nonVoting = value - voting;
+    uint256 penalty = nonVoting;
+    uint256 reward = penalty / 2;
+    electionSlashTest.vote(group, voting, address(0), address(0));
+    helper_WhenAccountIsSlashedForAllOfItsLockedGold(penalty, reward);
+
+    assertEq(lockedGold.getAccountTotalLockedGold(address(this)), value - penalty);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: ditto

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 23]
```diff
@@ -0,0 +1,2890 @@
...
+    uint256 voting = value / 2;
+    uint256 penalty = value;
+    uint256 reward = penalty / 2;
+    electionSlashTest.vote(group, voting, address(0), address(0));
+    helper_WhenAccountIsSlashedForAllOfItsLockedGold(penalty, reward);
+
+    assertEq(address(governance).balance, penalty - reward);
+  }
+
+  function test_ShouldSlashWholeAccountBalance_WhenAccountIsSlashedForMoreThanItsOwnBalance_WhenTheAccountHasHalfVotingAndHalfNonVotingGold()
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:1262]: is this case possible in practice or it's just in case there's an edge case? I'd clarify with a comment

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 24]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+
+  function test_ShouldRevertWhenDelegatingMoreThan100PercentInTwoStepsForDifferentDelegatees_WhenDelegatingToDelegatee1_WhenSomeGoldIsLocked()
+    public
+  {
+    uint256 value = 1000;
+
+    vm.prank(delegator);
+    lockedGold.lock.value(value)();
+    lockedGold.delegateGovernanceVotes(delegatee1, FixidityLib.newFixedFraction(10, 100).unwrap());
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I personally find it much easier to just write `0.1e24`.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 25]
```diff
@@ -0,0 +1,2890 @@
...
+
+    createAndAssertDelegatorDelegateeSigners(
+      delegator2,
+      delegatee2,
+      delegatorSigner2PK,
+      delegateeSigner2PK
+    );
+  }
+
+  function test_ShouldRevertWhenIncorrectPercentAmountIsInserted_WhenSomeGoldIsLocked_WhenVoteSigners()
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:1572]: would clarify what WhenVoteSigners means

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 26]
```diff
@@ -0,0 +1,2890 @@
...
+    lockedGold.delegateGovernanceVotes(delegatee2, FixidityLib.newFixedFraction(40, 100).unwrap());
+
+    address[] memory delegateesOfDelegator = lockedGold.getDelegateesOfDelegator(delegator);
+    assertEq(delegateesOfDelegator.length, 2);
+    assertEq(delegateesOfDelegator[0], delegatee1);
+    assertEq(delegateesOfDelegator[1], delegatee2);
+    vm.stopPrank();
+    lockedGold.setMaxDelegateesCount(3);
+    vm.prank(delegator);
+    lockedGold.delegateGovernanceVotes(delegatee3, FixidityLib.newFixedFraction(10, 100).unwrap());
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:1740]: worth checking that the delegation was correct.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 27]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+
+  function test_ShouldRevertWhenTryingToAddExtraDelegatee_WhenVoteSigners() public {
+    helper_WhenVoteSigners();
+    lockedGold.setMaxDelegateesCount(2);
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Can you replace the body of this function for `test_ShouldRevertWhenTryingToAddExtraDelegatee`?

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 28]
```diff
@@ -0,0 +1,2890 @@
...
+    address[] memory delegateesOfDelegator = lockedGold.getDelegateesOfDelegator(delegator);
+    assertEq(delegateesOfDelegator.length, 2);
+    assertEq(delegateesOfDelegator[0], delegatee1);
+    assertEq(delegateesOfDelegator[1], delegatee2);
+
+    vm.expectRevert("Too many delegatees");
+    lockedGold.delegateGovernanceVotes(delegatee3, FixidityLib.newFixedFraction(1, 1).unwrap());
+  }
+
+  function test_ShouldAllowToAddExtraDelegatee_WhenLimitIsIncreased_WhenVoteSigners() public {
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:1747]: Ditto with `test_ShouldAllowToAddExtraDelegatee_WhenLimitIsIncreased`

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 29]
```diff
@@ -0,0 +1,2890 @@
...
+      delegatedAmount + amountFromDelegator1AfterRevoke
+    );
+    assertEq(
+      governance.removeVotesCalledFor(delegatee1),
+      delegatedAmount + amountFromDelegator1AfterRevoke
+    );
+
+  }
+
+  function test_ShouldRevokeVotesCorrectlyWhenDelegateeVoting_WhenRevokingPercentageSuchAsThatWithNewlyLockedAmountIwWouldNotDecreaseBelowZero()
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I'd create a new class for the `WhenRevokingPercentageSuchAsThatWithNewlyLockedAmountIwWouldNotDecreaseBelowZero`

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 30]
```diff
@@ -0,0 +1,2890 @@
...
+    uint256 votingAmount = (delegatedAmount * 2 - amountToRevoke);
+
+    uint256 amountFromDelegator1AfterRevoke = ((2 * delegatedAmount) / percentageToDelegate) *
+      (percentageToDelegate - percentageToRevoke - percentageToRevokeAfterLock);
+
+    vm.prank(delegator);
+    lockedGold.lock.value(value)();
+    vm.prank(delegator2);
+    lockedGold.lock.value(value)();
+
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Didn't check exactly but this body also looks repeated.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 31]
```diff
@@ -0,0 +1,2890 @@
...
+  }
+
+  function test_ShouldReturnCorrectValueWhenLockedAndDelegatedForDelegateeAndDelegator_WhenOnlyDelegated_WhenHavingAccounts()
+    public
+  {
+    uint256 delegatedPercent = 70;
+    uint256 delegatedAmount = (value / 100) * delegatedPercent;
+
+    vm.prank(delegator);
+    accounts.createAccount();
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:2136]: it wouldn't hurt to create the accounts in the setUp

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 32]
```diff
@@ -0,0 +1,2890 @@
...
+      FixidityLib.newFixedFraction(percent, 100).unwrap()
+    );
+
+    (uint256 expectedAmount, uint256 realAmount) = lockedGold
+      .getDelegatorDelegateeExpectedAndRealAmount(delegator, delegatee);
+    assertEq(expectedAmount, amount);
+    assertEq(realAmount, amount);
+  }
+
+  function test_ShouldReturnEqualAmountAndUpdateTotalVotingPowerOfDelegatee_WhenMoreCeloLocked_WhenVoteSigners()
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:2318]: ditto here, most of the body looks repeated

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 33]
```diff
@@ -0,0 +1,2890 @@
...
+    vm.deal(delegatee, 10 ether);
+
+    vm.prank(delegator);
+    lockedGold.lock.value(value)();
+
+    (delegatorSigner, delegatorSignerPK) = actorWithPK("delegatorSigner");
+    (delegateeSigner, delegateeSignerPK) = actorWithPK("delegateeSigner");
+  }
+
+  function helper_WhenVoteSigners() public {
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: This function is also repeated and could go to the parent or an abstract contract.

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol [hunk 34]
```diff
@@ -0,0 +1,2890 @@
...
+    lockedGold.updateDelegatedAmount(delegator, delegatee);
+
+    assertEq(lockedGold.getAccountTotalLockedGold(delegator), value * 2);
+    assertEq(lockedGold.getAccountTotalGovernanceVotingPower(delegatee), delegatedAmount * 2);
+    assertDelegatorDelegateeAmounts(delegator, delegatee, delegatedPercent, delegatedAmount * 2);
+  }
+}
+
+contract GetTotalPendingWithdrawalsCount is LockedGoldFoundryTest {
+  function setUp() public {
```
Comment (martinvol) [inline packages/protocol/test-sol/governance/voting/LockedGold.t.sol:2408]: Super can be removed

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 9 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I added caller variable, but I dont believe we can add it as constant since it is dynamic value

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 13 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Most of our tests is designed as unit tests - meaning lots of interacting smart contracts are mocked. We could change our testing to integration tests/ e2e tests but it would require significant rewrite

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 24 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I created helper functions that hopefully make it more readable

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 10 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol:321]: I would prefer to keep randomAddress since from msg.sender is not clear what is actually being tested

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 14 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol:501]: yep, GitHub seems to have bug

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 15 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol:510]: I believe this test and previous test should be sufficient

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 23 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol:1262]: I believe lots of tests are in general just edge cases - Im not sure whether the comment would be very helpfull

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 27 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: In general I did lots of refactoring and I was able to remove 25% of code

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 21 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: We dont have method to check voting gold - it is necessary to get totalLockedGold and subtract nonVotingGold  (please note that this is Election voting rather then Governance voting)

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 16 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: Account is not delegating

File: packages/protocol/test-sol/governance/voting/LockedGold.t.sol (hunk 6 above)
Comment (pahor167) [reply packages/protocol/test-sol/governance/voting/LockedGold.t.sol]: I updated test names so it is more clear
